controllers/tokenupload_controller.go

```diff
@@ -0,0 +1,235 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/logs"
+
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/spi-shared/tokenstorage"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	spi "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const tokenSecretLabel = "spi.appstudio.redhat.com/upload-secret"
+const spiTokenNameLabel = "spi.appstudio.redhat.com/token-name"
+const providerUrlLabel = "spi.appstudio.redhat.com/providerUrl"
+
+// TokenUploadReconciler reconciles a Secret object
+type TokenUploadReconciler struct {
+	client.Client
+	Scheme       *runtime.Scheme
+	TokenStorage tokenstorage.NotifyingTokenStorage
+}
+
+func (r *TokenUploadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	lg := log.FromContext(ctx)
+
+	list := &corev1.SecretList{}
+
+	// list secrets labeled with "spi.appstudio.redhat.com/upload-secret"
+	err := r.List(ctx, list, client.HasLabels{tokenSecretLabel})
+	if err != nil {
+		lg.Error(err, "can not get list of secrets")
+		return ctrl.Result{}, fmt.Errorf("can not get list of secrets: %w,", err)
+	}
+
+	for i, s := range list.Items {
+
+		var accessToken spi.SPIAccessToken
+
+		// we immediatelly delete the Secret
+		err := r.Delete(ctx, &list.Items[i])
+		if err != nil {
+			logError(ctx, s, fmt.Errorf("can not delete the Secret: %w", err), r, lg)
+			continue
+		}
+
+		// if spiTokenName field is not empty - try to find SPIAccessToken by it
+		if s.Labels[spiTokenNameLabel] != "" {
+			spiTokenName := s.Labels[spiTokenNameLabel]
+			accessToken = spi.SPIAccessToken{}
+			err = r.Get(ctx, types.NamespacedName{Name: spiTokenName, Namespace: s.Namespace}, &accessToken)
+
+			if err != nil {
+				logError(ctx, s, fmt.Errorf("can not find SPI access token %s: %w ", spiTokenName, err), r, lg)
+				continue
+			} else {
+				lg.V(logs.DebugLevel).Info("SPI Access Token found : " + accessToken.Name)
+			}
+
+			// spiTokenName field is empty
+			// check providerUrl field and if not empty - try to find the token for this provider instance
+		} else if s.Labels[providerUrlLabel] != "" {
+			providerUrl := s.Labels[providerUrlLabel]
+			// NOTE: it does not fit advanced policy of matching token!
+			// Do we need it as an SPI "API function" which take into account this policy?
+			tkn := findTokenByUrl(ctx, providerUrl, s.Namespace, r, lg)
```

Comment:
should we somehow validate the url so it's in format we need?

Reply:
we do not use it as URL in the function, it is just a string to compare, so we can add one more validation to have a nice message in a case URL is not valid but it hardly bring any value IMO because the fact that it is not found (which should be the case b/c I suppose all Token's URLs are valid.